File: problems/1652_defuse_the_bomb/csharp/1652_defuse_the_bomb.cs
using System.Linq;
using System;

public static class Program
{
  public static void Main(string[] args)
  {
    Solution s = new Solution();
    Console.WriteLine(string.Join(", ", s.Decrypt(new int[] { 5, 7, 1, 4 }, 3)));
    Console.WriteLine(string.Join(", ", s.Decrypt(new int[] { 1, 2, 3, 4 }, 0)));
    Console.WriteLine(string.Join(", ", s.Decrypt(new int[] { 2, 4, 9, 3 }, -2)));
  }
}

public class Solution
{
  int LoopIdx(int idx, int len) { return (idx + len) % len; }
  public int[] Decrypt(int[] code, int k)
  {
    int len = code.Length;
    int[] result = code.Select(x => 0).ToArray();

    if (k == 0)
      return result;

    if (k < 0)
      Array.Reverse(code);

    for (int i = 0; i < len; i++)
      result[i] = SumK(code, i, Math.Abs(k));

    if (k < 0)
      Array.Reverse(result);

    return result;
  }
  int SumK(int[] code, int i, int k)
  {
    int result = 0;

    for (int j = 0; j < k; j++)
      result += code[LoopIdx(i + 1 + j, code.Length)];

    return result;
  }
}

File: problems/1652-defuse-the-bomb/python/1652_defuse_the_bomb.py
from typing import List, Self


class Solution:
    def decrypt(self: Self, code: List[int], k: int) -> List[int]:
        length: int = len(code)
        result: List[int] = [0 for _ in code]

        if k == 0:
            return result

        if k < 0:
            code.reverse()

        for i in range(length):
            result[i] = self.sumK(code, i, abs(k))

        if k < 0:
            result.reverse()

        return result

    def loopIdx(self: Self, idx: int, len: int) -> int:
        return (idx + len) % len

    def sumK(self: Self, code: List[int], i: int, k: int) -> int:
        result: int = 0

        for j in range(k):
            result += code[self.loopIdx(i + 1 + j, len(code))]

        return result


def test(code: List[int], k: int) -> None:
    result: List[int] = Solution().decrypt(code, k)
    print(f"Decrypted code for this code and k is {
          result}! \n -> code: {code} \n -> k: {k}")


if __name__ == "__main__":
    test([5, 7, 1, 4], 3)
    test([1, 2, 3, 4], 0)
    test([2, 4, 9, 3], -2)
